go/vt/topotools/routing_rules.go

```diff
@@ -158,4 +166,33 @@ func SaveKeyspaceRoutingRules(ctx context.Context, ts *topo.Server, rules map[st
 	return ts.SaveKeyspaceRoutingRules(ctx, keyspaceRoutingRules)
 }
 
+// UpdateKeyspaceRoutingRules updates the keyspace routing rules in the topo server. It initially acquires a
+// RoutingRulesLock and then calls the update function to modify the rules in-place.
+// If the update function returns an error, the rules are not saved and the lock is released.
+// If the update function is successful, the rules are saved to the topo and the lock is released.
+func UpdateKeyspaceRoutingRules(ctx context.Context, ts *topo.Server, reason string,
+	update func(ctx context.Context, rules *map[string]string) error) (err error) {
+	var lock *topo.RoutingRulesLock
+	lock, err = topo.NewRoutingRulesLock(ctx, ts, reason)
+	if err != nil {
+		return err
+	}
+	lockCtx, unlock, lockErr := lock.Lock(ctx)
+	if lockErr != nil {
+		return lockErr
+	}
+	defer unlock(&err)
+	rules, _ := GetKeyspaceRoutingRules(lockCtx, ts)
+	if rules == nil {
+		rules = make(map[string]string)
+	}
+	if err := update(lockCtx, &rules); err != nil {
+		return err
+	}
+	if err := saveKeyspaceRoutingRulesLocked(lockCtx, ts, rules); err != nil {
+		return err
+	}
+	return nil
```

Comment:
IMO this is much nicer and it eliminates all the awkward sentinel related code, while remaining more standard for the topo server, and most importantly ensuring that we're only ever making the minimum necessary topo calls:
```
	update func(ctx context.Context, rules *map[string]string) error) (err error) {
	var lock *topo.RoutingRulesLock
	lock, err = topo.NewRoutingRulesLock(ctx, ts, reason)
	if err != nil {
		return err
	}
	lockCtx, unlock, lockErr := lock.Lock(ctx)
	if lockErr != nil {
		// If the key does not yet exist then let's create it.
		if !topo.IsErrType(lockErr, topo.NoNode) {
			return lockErr
		}
		rules := make(map[string]string)
		if err := update(lockCtx, &rules); err != nil {
			return err
		}
		// This will fail if the key already exists and thus avoids any races here.
		if err := ts.CreateKeyspaceRoutingRules(ctx, buildKeyspaceRoutingRules(&rules)); err != nil {
			return err
		}
		return nil
	}
	defer unlock(&err)
	rules, err := GetKeyspaceRoutingRules(lockCtx, ts)
	if err != nil {
		return err
	}
	if err := update(lockCtx, &rules); err != nil {
		return err
	}
	if err := saveKeyspaceRoutingRulesLocked(lockCtx, ts, rules); err != nil {
		return err
	}
	return nil
```

Full diff (including updated unit tests): https://gist.github.com/mattlord/68e8636fe65cf72baab31e8f6b7b8604

If you disagree, can you help me understand why?

Reply:
> So if I understand correctly your issues with current code are:
> 
> * There is that extra `sentinel` key that stays around
> * We have an extra topo `get` for every lock

It's more than that too: https://github.com/vitessio/vitess/pull/15807#discussion_r1597636396

> With the recommended code:
> 
> * There is no sentinel and extra check

That's trivializing it a bit IMO, just as it would be to ignore us doing something very unusual with MySQL that breaks well established usage/access patterns. And doing so unnecessarily, and in a way that is slower / more costly. 

We could, of course, alternatively require that you create the rules before updating them in a more direct way  (`CreateKeyspaceRoutingRules`). That would be even more standard — in that model we would always check for `topo.NoNode` errors and do `CreateKeyspaceRoutingRules` instead (as we're doing automatically in my suggestion).

> * The caller can (most likely very rarely) need to retry. So every caller of the `Update` needs to be aware of the need to retry or pass it to the the client who needs to know that this is not an error but need to retry).

The caller (the vitess function or the vitess user) can always get an error and can retry. For `topo.NodeExists` in the case of an update, the caller would only ever *at most* get this once and retry an update because of it once — as if you get that error on update you know that it now exists and you can of course then update it. And we don't ever allow deleting it (although someone could do that directly in the underlying topo implementation like `etcd` of course) so it's generally a one time scenario.

I added the auto retry on `topo.NodeExist` error in the create client RPC and the workflow utils update function as those looked like the two places where you'd likely be creating the key (the utils update function is used when creating the initial routing rules for a workflow). IMO, no other callers / locations need to worry about `topo.NodeExists` errors.

> I guess in your approach we can add N retries instead of the one and make the chance of the race surfacing the user practically zero.

No need to retry more than once. Or retry at all really. Since only one caller should ever get the noted error (from then on the node will exist) in the lifetime of the topo server instance.

I don't understand under what scenario you could possibly race further here. There will always be one writer that wins (etcd2topo and etcd does this `Create`->`Put` in a strictly serializable transaction).

> In fact, I did start with an approach similary to yours. My thought was originally that since the topo (at least etcd) can lock a keypath without the key needing to exist, I could obtain a lock first and then create the rules. Unfortunately for this approach, Vitess requires that the key path be a "dir" and you cannot lock leaf keys.

Yeah, I think that these issues are in part arising because `RoutingRules` was a later topo addition and not done in a standard way (with a path, create, locking, etc). I'm hoping that we can avoid doing more non-standard things here.

> I didn't think it was an option to surface a race to the end user, so came up with the sentinel approach. Seemed like Vitess should manage it, because it can.

`topo.NoNode` and `topo.NodeExists` are not unusual to surface. If you try to update things in the topo before they exist, you get NoNode (that's why most things use create/update vs apply for the others that always do a blind write). And if you try to create something that already exists you get NodeExists. These are not unusual to have surface in the UI. 

And my suggestion *does* prevent this from surfacing to the user, just as the sentinel file method did. 

> Regarding the extra `get`, a similar approach of trying to lock first and creating only on a `NoNode` can remove the extra `get`s. But the sentinel key will indeed remain.
> 
> Another point to note is that we delete the routing rules files today if there are no rules. So we would be creating the rules more than just once.

Why would we delete the file/key? IMO we should only delete the rules we want to delete as empty/no rules is not the same as having no key. Think of it like a table with no rows vs no table. That would be yet another non-standard thing IMO that I must have missed. Where do we do that? Ah, I see:
```
if len(data) == 0 {
		// No rules, remove it.
		if err := ts.globalCell.Delete(ctx, ts.GetKeyspaceRoutingRulesPath(), nil); err != nil && !IsErrType(err, NoNode) {
```

There's no reason to do that. IMO we should simply write the empty rules. You're saying that you want to apply empty routing rules, not that you want to delete the file. This IMO is like turning a `delete from tb1` into a `drop table t1`.

> If there are always multiple migrations in flight we will create it once, but if the user runs set of migrations and waits for them to complete before starting the next set, we will again follow the flow where the key to be locked needs to be created. With the sentinel file the prefix being locked is always present.

I think that this is resulting from another unusual and unnecessary behavior as noted above. There's no need to delete the file, just the rules within it (just as you don't delete the global routing rules file/key -- there's no command for it).

> I think it is more important to get this feature in, to add safety to ongoing migrations. So @deepthi and @mattlord, if you are both fine with @mattlord's suggested approach, please go ahead and push that change to this PR, so that the attribution is correct and we minmise iterations. I can review the changes once they are in, just as an extra 👀

I could push my suggestions if you want. The diff also clearly lays them them out. 

I'll remove my `require changes` block on the PR now that we've at least discussed things. Perhaps others are fine with things as-is and you can move forward with their review (I've spent A LOT of time on this to date). It could always be changed/improved later too. The topo paths would have been much more difficult to change later but I think we've resolved those aspects well. These other things we're discussing are more internal and could be changed in a non-user visible and upgrade/downgrade safe way later.

Reply:
If you do end up moving forward with things as they are, I think that this is an important point to address: https://github.com/vitessio/vitess/pull/15807/files#r1597650914

Reply:
I was in fact somewhat uncomfortable with adding a sentinel file. If we can indeed work out a way to avoid that, that will be better.

Reply:
Regarding deleting the rules file @rohit-nayak-ps do we currently do that with Table/Shard routing rules as well?
As for why they may be different from other entities we store in the topo, I can only guess that it is because unlike keyspaces/shards/tablets, routing rules were intended to be temporary in the first place.

Reply:
I cannot say for sure why things related to the routing rules were done in a certain way in the past, but it does look like they were initially added here AFAICT: https://github.com/vitessio/vitess/pull/4833/files#diff-c12ed500f6b6ee0d413204f2a22a0885caf9daaa427cc636de912fd5a93e0320

I can, however, say the following things, which are largely based on my own opinions and preferences so it's perfectly fine if others disagree:
  1. The routing rules related topo work came ~ 2 years later than virtually all of the general topo server related work (~5 vs ~ 7 years). That general topo server design/work was largely done by Alain and later work was done by others (so different times, perspectives, pressures, etc).
  2. I think that the general design/work was well thought out and well done — to me it seems that time was clearly taken to write a well thought out design/architecture which was later implemented. Some of the later work was IMO done in a more rushed fashion as we were more/most concerned with trying to quickly achieve the needed given product/user feature — which is normal and fine, no judgement there.
  3. Lastly, and for me most importantly, if I reviewed a PR today where the author was dropping a sidecar table because there were no rows in it after the action being taken, I would object. It's the same fundamental principle for the topo data store in my mind.

Reply:
> // No vschema, remove it. So we can remove the keyspace.

That comment makes no sense to me in Vitess today FWIW.

My larger point is that the RoutingRules weren't really defined as a permanent part of the Vitess data model before — it was temporary storage for a blob so no rules (no pun intended, I mean no data store rules). Write a temporary blob here. 

But now we want to — at least in my mind — make `RoutingRules` a well defined part of the Vitess data model. And with that, apply the normal patterns, rules, and usage. We now want concurrency controls, consistency, etc. So we define a container to hold this specific type of element with a specific structure (it's like a table in our sidecar database). There's no reason to drop/delete that container whenever it happens to be empty. That would be wrong IMO in principle, let alone that it causes far more unnecessary work and slows system performance.

One could argue that no, we should drop it and recreate it when actually needed because in their opinion the data model should only be defined in the application and not the data store. OK, I disagree but that's a valid opinion. But to then say, instead of keeping the container for the type permanently in our data store, we're going to keep some other random thing (the sentinel file) permanently in the data store so that we can get around the rules and access patterns already defined for containers in the data store — I think that is just wrong.

The larger context in my mind is that the `VSchema` — routing rules becoming a part of that in `VTGate` — are a critical part of Vitess. A critical part of our data model and our operating model. So we should start treating them that way. Making them a more standard part of the data model so that we can ensure proper / expected behavior with proper concurrency controls, consistency guarantees, versioning etc is pretty important in my mind. I'm trying to push us more methodically in that direction, starting here, so that we can eventually properly address issues like this: https://github.com/vitessio/vitess/issues/15794

I personally think this is important and it's why I took the time to show exactly — with a working diff — how we could move in that direction while still achieving the immediate larger feature goals of this PR in a timely manner.